zingo-testutils/src/lib.rs

```diff
@@ -243,6 +244,272 @@ pub async fn load_wallet(
+        for utxo in spent_utxos {
+            if utxo["spent"] == txid || utxo["unconfirmed_spent"] == txid {
+                tx_spend += utxo["value"].as_u64().unwrap_or(0);
+            }
+        }
+    }
+    if let JsonValue::Array(pending_utxos) = &notes["pending_utxos"] {
+        for utxo in pending_utxos {
+            if utxo["spent"] == txid || utxo["unconfirmed_spent"] == txid {
+                tx_spend += utxo["value"].as_u64().unwrap_or(0);
+            } else if utxo["created_in_txid"] == txid {
+                tx_change += utxo["value"].as_u64().unwrap_or(0);
+            }
+        }
+    }
+    if let JsonValue::Array(unspent_utxos) = &notes["utxos"] {
+        for utxo in unspent_utxos {
+            if utxo["created_in_txid"] == txid {
+                tx_change += utxo["value"].as_u64().unwrap_or(0);
+            }
+        }
+    }
+
+    if let JsonValue::Array(spent_sapling_notes) = &notes["spent_sapling_notes"] {
+        for note in spent_sapling_notes {
+            if note["spent"] == txid || note["unconfirmed_spent"] == txid {
+                tx_spend += note["value"].as_u64().unwrap_or(0);
+            }
+        }
+    }
+    if let JsonValue::Array(pending_sapling_notes) = &notes["pending_sapling_notes"] {
+        for note in pending_sapling_notes {
+            if note["spent"] == txid || note["unconfirmed_spent"] == txid {
+                tx_spend += note["value"].as_u64().unwrap_or(0);
+            } else if note["created_in_txid"] == txid {
+                tx_change += note["value"].as_u64().unwrap_or(0);
+            }
+        }
+    }
+    if let JsonValue::Array(unspent_sapling_notes) = &notes["unspent_sapling_notes"] {
+        for note in unspent_sapling_notes {
+            if note["created_in_txid"] == txid {
+                tx_change += note["value"].as_u64().unwrap_or(0);
+            }
+        }
+    }
+
+    if let JsonValue::Array(spent_orchard_notes) = &notes["spent_orchard_notes"] {
+        for note in spent_orchard_notes {
+            if note["spent"] == txid || note["unconfirmed_spent"] == txid {
+                tx_spend += note["value"].as_u64().unwrap_or(0);
+            }
+        }
+    }
+    if let JsonValue::Array(pending_orchard_notes) = &notes["pending_orchard_notes"] {
+        for note in pending_orchard_notes {
+            if note["spent"] == txid || note["unconfirmed_spent"] == txid {
+                tx_spend += note["value"].as_u64().unwrap_or(0);
+            } else if note["created_in_txid"] == txid {
+                tx_change += note["value"].as_u64().unwrap_or(0);
+            }
+        }
+    }
+    if let JsonValue::Array(unspent_orchard_notes) = &notes["unspent_orchard_notes"] {
+        for note in unspent_orchard_notes {
+            if note["created_in_txid"] == txid {
+                tx_change += note["value"].as_u64().unwrap_or(0);
+            }
+        }
+    }
+
+    tx_spend - tx_change
+}
```

Comment:
this function seems to return the total value transferred from a wallet which equals the amount sent + fee. for the zip317 test framework we are concerned with the fee only. Maybe we can determine the fee by calling this function on the recipient aswell but I feel a more elegant solution is to check the actual note structs within the transactions in the wallet and determine all information from the sender alone.

Reply:
i dont think its possible to calculate fees solely from notes known by the sending wallet, this is the reason ive been using total transfer value and subtracting the amount sent in do_send, but happy to make changes if you can suggest the code that will do this.